packages/contracts/test/UnirepApp.test.ts

```diff
@@ -55,12 +61,86 @@ describe('Unirep App', function () {
+            .to.emit(app, 'UserInitSuccess')
+            .withArgs(hashUserId)
+    })
+
+    it('user sign up after init', async () => {
+        const userState = await genUserState(id, app)
+
+        // get user status, should be RegisterStatus.INIT
+        expect(await app.queryUserStatus(hashUserId)).to.equal(1)
+
+        const { publicSignals, proof } = await userState.genUserSignUpProof()
+        expect(await app.userSignUp(publicSignals, proof, hashUserId, false))
+            .to.emit(app, 'UserSignUpSuccess')
+            .withArgs(hashUserId)
+
+        // get user status, should be RegisterStatus.REGISTERED_SERVER
+        expect(await app.queryUserStatus(hashUserId)).to.equal(2)
+
+        userState.stop()
+    })
+
+    it('revert when user sign up multiple times', async () => {
+        const userState = await genUserState(id, app)
+
+        const { publicSignals, proof } = await userState.genUserSignUpProof()
+        expect(
+            app.userSignUp(publicSignals, proof, hashUserId, false)
+        ).to.be.revertedWithCustomError(app, 'UserAlreadySignedUp')
+
+        userState.stop()
+    })
+
+    it('revert when user sign up without init', async () => {
+        // create another random user id
+        const id = new Identity()
+        const hash = crypto.createHash('sha3-224')
+        const hashUserId = `0x${hash
+            .update(id.getCommitment().toString())
+            .digest('hex')}`
+
+        const userState = await genUserState(id, app)
+
+        // get user status, should be RegisterStatus.NOT_REGISTER
+        expect(await app.queryUserStatus(hashUserId)).to.equal(0)
+
+        const { publicSignals, proof } = await userState.genUserSignUpProof()
+        expect(
+            app.userSignUp(publicSignals, proof, hashUserId, true)
+        ).to.be.revertedWithCustomError(app, 'UserInitStatusInvalid')
+
+        userState.stop()
+    })
+
+    it('revert when reuse user signup proof', async () => {
+        // create another random user id
+        const id = new Identity()
+        const hash = crypto.createHash('sha3-224')
+        const hashUserId = `0x${hash
+            .update(id.getCommitment().toString())
+            .digest('hex')}`
+
         const userState = await genUserState(id, app)
 
-        // generate
+        expect(await app.initUserStatus(hashUserId))
+            .to.emit(app, 'UserInitSuccess')
+            .withArgs(hashUserId)
+
         const { publicSignals, proof } = await userState.genUserSignUpProof()
-        await app.userSignUp(publicSignals, proof).then((t) => t.wait())
+        const invalidProof = proof
+            .slice(0, proof.length - 1)
+            .concat([BigInt(0)])
+
+        // get user status, should be RegisterStatus.INIT
+        expect(await app.queryUserStatus(hashUserId)).to.equal(1)
+
+        expect(app.userSignUp(publicSignals, invalidProof, hashUserId, true)).to
+            .be.reverted;
+
```

Comment:
We could use a `describe()` hook to contain all the sign up tests, which would be neater.